design-documents/services-orchestration.md

```diff
@@ -0,0 +1,254 @@
+    </choice>
+</step>
+```
+The conditional operation allows to branch execution logic by specifying a set of particular conditions per start of each new branch.
+The conditional step supports complex nested condition which can be use `AND`, `OR`, `NOR`.
+## Error handling
+With catch node, it is possible to declare a handler for an exceptional case. Each exception may have its own handler.
+Also, it is possible to declare retry operation, the declaration is similar to exception handling declaration.
+The trigger for retry is an exception.
+```xml
+<step xsi:type="Task" name="ExecuteStep4" next="Exit" routine="example:step:2">
+    <catch exceptoin="InvalidArgumentException" next="Handler" />
+    <retry exceptoin="ServiceUnavailableException" attempts="3" timeout="100" />
+</step>
+```
+
+### Operation terminators (Success/Failure)
+`<step name="ScenarioSuccessfullyCompleted" xsi:type="Success" outputPath="params"/>`
+
+### Trivial Example
+[Configuration schema](https://github.com/akaplya/mage-state-machine/blob/master/lib/internal/Magento/Framework/StateMachine/etc/stateMachine.xsd)
+
+```xml
+    <scenario name="ExampleRun">
+        <step xsi:type="Task" name="ExecuteStep1" next="ExecuteStep2"
+              inputPath="params/foo"
+              outputPath="params/bar"
+              routine="example:step:1"
+        />
+        <step xsi:type="Task"
+              name="ExecuteStep2"
+              next="ScenarioSuccessfullyCompleted"
+              inputPath="params/bar"
+              routine="example:step:2"
+        />
+        <step name="ScenarioSuccessfullyCompleted" xsi:type="Success" outputPath="params"/>
+    </scenario>
+```
+
+## Routines
+Routines configuration specifies a callable resource to be used in scenario steps.
+
+Actually we have to support two base types of routines:
+* Direct PHP Calls.
+* Magento WebAPI  REST calls.
+
+### Routines declaration example
+```
+    <routine xsi:type="MagentoWebAPIRoutine"
+            name="Magento:Example:Step1"
+            method="PUT"
+            url="http:/magento2sm.local/rest/V1/example/step1"
+    />
+    <routine xsi:type="MagentoWebAPIRoutine"
+             name="Magento:Example:Step2"
+             method="PUT"
+             url="http:/magento2sm.local/rest/V1/example/step2"
+    />
+``` 
+
+[Configuration schema](https://github.com/akaplya/mage-state-machine/blob/master/lib/internal/Magento/Framework/StateMachine/etc/routines.xsd)
+
+## Backward compatibility
+
+Such a mechanism can be useful with the evolution of an existing application.
+Major scenarios can be described with such language.
+For monolithic app the scenario rewrote such way will be backward compatible because scenario will run the same operation in the monolith execution context.
+### Real life example
+
+Implementation of [\Magento\Sales\Api\OrderManagementInterface::place](https://github.com/magento/magento2ce/blob/2.3-develop/app/code/Magento/Sales/Model/Service/OrderService.php#L196) 
```

Comment:
Where this configuration would live?

Reply:
Modules and configuration decomposition is aligned with the base document. So the main scenario should be declared in the scope of the corresponding module group. For instance `place order` scenario will be declared in SalesWorkflow. Payment extension will come from SalesPaymentWorkflow module.

Reply:
Will this module be deployed on a orchestration service?